### PR TITLE
Head tail operators

### PR DESCRIPTION
**Pull request by ahmet-uyar**
I have implemented head and tail operators for gcylon DataFrame
I also implemented row_counts_for_all method. Users can query the row_counts of all partitions. Row counts are all-gathered to all workers. We use this method when calculating head/tail rows counts in workers. 

Head/Tail methods construct a new DataFrame with the given row_counts. 
For the head method: 
* If the first worker has more than n rows,
  all rows will be returned from the first worker.
  Head n rows of the first worker will be returned.
* If the first worker has less than n rows, then:
  rows from the second worker will be returned.
* If the second worker also does not have enough rows,
  rows from the third worker will be returned.
  etc.

Tail method is similar. It performs these calculations starting from the last worker. 

I have added examples and tests cases.

### CODE REVIEW

File: cpp/src/examples/gcylon/gsorting_plus.cpp [hunk 1]
```diff
@@ -0,0 +1,226 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (supunkamburugamuve) [inline cpp/src/examples/gcylon/gsorting_plus.cpp:2]: I suggest not having this file. It seems like a test case

**Review by supunkamburugamuve — COMMENTED** [review]
In general we use 2 spaces for tabs.

File: cpp/src/examples/gcylon/gsorting_plus.cpp (hunk 1 above)
Comment (ahmet-uyar) [reply cpp/src/examples/gcylon/gsorting_plus.cpp]: removed.

File: python/pygcylon/examples/row_counts_all.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    cdf = util.random_data_df(nrows=20 + env.rank, col_lows=[0, 100, 200, 300], col_highs=[50, 150, 250, 350])
+    cdf = cdf.set_index(keys=["col-0", "col-3"])
+    print("\n\nnumber of rows: ", len(cdf))
+
+    gdf = gcy.DataFrame(cdf)
+    row_counts = gdf.row_counts_for_all(env=env)
+    print("\n\nnumber of rows in all dataframe partitions: ", row_counts)
+    if isinstance(row_counts, list):
+        print("row_counts is a list. len: ", len(row_counts))
+    else:
```
[the remark this answers is not in the document]
Comment (ahmet-uyar) [reply python/pygcylon/examples/row_counts_all.py]: done.

File: cpp/src/cylon/net/ops/gather.cpp [hunk 3]
```diff
@@ -153,4 +153,24 @@ cylon::Status cylon::mpi::Gather(const std::shared_ptr<cylon::TableSerializer> &
     return cylon::Status::OK();
 }
 
+cylon::Status cylon::mpi::AllGather(const std::vector<int32_t> &send_data,
+                                    int number_of_workers,
```
[the remark this answers is not in the document]
Comment (ahmet-uyar) [reply cpp/src/cylon/net/ops/gather.cpp:157]: in all gcylon, the indentation is 3. so probably i need to fix that separately.

**Review by supunkamburugamuve — APPROVED** [review]

**Pull request merged**